lib/Common/Codex/Utf8Helper.h

```diff
@@ -70,14 +69,44 @@ namespace utf8
             return E_OUTOFMEMORY;
         }
 
-        // Some node tests depend on the utf8 decoder not swallowing invalid unicode characters
-        // instead of replacing them with the "replacement" chracter. Pass a flag to our
-        // decoder to require such behavior
-        utf8::DecodeUnitsIntoAndNullTerminateNoAdvance(destString, (LPCUTF8) sourceString, (LPCUTF8) sourceString + cbSourceString, DecodeOptions::doAllowInvalidWCHARs);
-        Assert(destString[cchDestString] == 0);
-        static_assert(sizeof(utf8char_t) == sizeof(char), "Needs to be valid for cast");
-        *destStringPtr = destString;
-        *destCount = cchDestString;
+        if (allocateCount != nullptr) *allocateCount = cbDestString;
+
+        for (; sourceStart < sourceCount; sourceStart++)
+        {
+            const char ch = sourceString[sourceStart];
+            if ( ! (ch > 0 && ch < 0x0080) )
+            {
+                int fallback = sourceStart > 3 ? 3 : sourceStart; // 3 + 1 -> fallback at least 1 unicode char
+                sourceStart -= fallback;
+                break;
+            }
+            destString[sourceStart] = (WCHAR) ch;
+        }
+
+        if (sourceStart == sourceCount)
+        {
+            *destCount = sourceCount;
+            destString[sourceCount] = WCHAR(0);
+            *destStringPtr = destString;
+        }
+        else
+        {
+            LPCUTF8 remSourceString = (LPCUTF8)sourceString + sourceStart;
+            WCHAR *remDestString = destString + sourceStart;
+
+            charcount_t cchDestString = utf8::ByteIndexIntoCharacterIndex(remSourceString, cbSourceString - sourceStart);
+            cchDestString += sourceStart;
+            Assert (cchDestString <= sourceCount);
+
+            // Some node tests depend on the utf8 decoder not swallowing invalid unicode characters
+            // instead of replacing them with the "replacement" chracter. Pass a flag to our
+            // decoder to require such behavior
+            utf8::DecodeUnitsIntoAndNullTerminateNoAdvance(remDestString, remSourceString, (LPCUTF8) sourceString + cbSourceString, DecodeOptions::doAllowInvalidWCHARs);
```

Comment:
We send the end buffer ptr to `DecodeUnitsIntoAndNullTerminateNoAdvance` and there is an assert for `\0` check